sql/parser: panic: SET TIME ZONE INTERVAL 'string' MINUTE TO SECOND: interface conversion: interface is parser.sqlSymUnion, not parser.durationField

Found with RSG: https://teamcity.cockroachdb.com/viewLog.html?tab=buildLog&logTab=tree&filter=debug&expand=all&buildId=98768#_focus=688

```
panic: interface conversion: interface is parser.sqlSymUnion, not parser.durationField [recovered]
	panic: SET TIME ZONE INTERVAL 'string' MINUTE TO SECOND: interface conversion: interface is parser.sqlSymUnion, not parser.durationField [recovered]
	panic: SET TIME ZONE INTERVAL 'string' MINUTE TO SECOND: interface conversion: interface is parser.sqlSymUnion, not parser.durationField

goroutine 22236 [running]:
panic(0x18705a0, 0xc420ba13d0)
	/usr/local/go/src/runtime/panic.go:500 +0x1a1
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc4200de000)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:185 +0x6e
panic(0x18705a0, 0xc420ba13d0)
	/usr/local/go/src/runtime/panic.go:458 +0x243
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).ExecuteStatements.func1(0xc4221be6d9, 0x30)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:455 +0x149
panic(0x18a1520, 0xc42234acc0)
	/usr/local/go/src/runtime/panic.go:458 +0x243
github.com/cockroachdb/cockroach/pkg/sql/parser.(*sqlParserImpl).Parse(0xc4222fc9b0, 0x2664300, 0xc4222fc8d8, 0x0)
	sql.y:1465 +0xc24f
github.com/cockroachdb/cockroach/pkg/sql/parser.(*Parser).Parse(0xc4222fc8d8, 0xc4221be6d9, 0x30, 0x0, 0xf, 0xc4221a10e8, 0x1, 0x1, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/parser/parse.go:71 +0x75
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).execRequest(0xc42024b1e0, 0xc4222fb600, 0xc4221be6d9, 0x30, 0x0, 0x0, 0x0, 0x0, 0xc4221a1300)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:526 +0x222
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).ExecuteStatements(0xc42024b1e0, 0xc4222fb600, 0xc4221be6d9, 0x30, 0x0, 0x0, 0x0, 0x0, 0xc4221be600)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:461 +0x11f
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*v3Conn).executeStatements(0xc420cfec00, 0x2b3cb4b3b080, 0xc42089cf30, 0xc4221be6d9, 0x30, 0x0, 0x0, 0x0, 0x0, 0xb48d01, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/v3.go:794 +0x94
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*v3Conn).handleSimpleQuery(0xc420cfec00, 0x2b3cb4b3b080, 0xc42089cf30, 0xc420cfec28, 0x35, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/v3.go:468 +0xb3
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*v3Conn).serve(0xc420cfec00, 0x2b3cb4b3b080, 0xc42089cf30, 0x5400, 0xc4200bc2f8, 0x2b3cb4b3b080, 0xc420e62420, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/v3.go:395 +0xa05
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*Server).ServeConn(0xc4200bc1c0, 0x2b3cb4b3b080, 0xc420e62420, 0x2678980, 0xc421266380, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/server.go:309 +0x999
github.com/cockroachdb/cockroach/pkg/server.(*Server).Start.func8.1(0x26798e0, 0xc4219af8c0)
	/go/src/github.com/cockroachdb/cockroach/pkg/server/server.go:522 +0x106
github.com/cockroachdb/cockroach/pkg/util/netutil.(*Server).ServeWith.func1(0xc4200de000, 0xc420174090, 0x26798e0, 0xc4219af8c0, 0xc42016bd40)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/netutil/net.go:136 +0x95
created by github.com/cockroachdb/cockroach/pkg/util/netutil.(*Server).ServeWith
	/go/src/github.com/cockroachdb/cockroach/pkg/util/netutil/net.go:138 +0x277
```

cc @mjibson

--- a/pkg/sql/parser/sql.go
+++ b/pkg/sql/parser/sql.go
@@ -8469,7 +8469,7 @@ sqldefault:
 		sqlDollar = sqlS[sqlpt-1 : sqlpt+1]
 		//line sql.y:3464
 		{
-			sqlVAL.union.val = sqlDollar[1].union
+			sqlVAL.union.val = sqlDollar[1].union.durationField()
 		}
 	case 564:
 		sqlDollar = sqlS[sqlpt-3 : sqlpt+1]
@@ -8493,7 +8493,7 @@ sqldefault:
 		sqlDollar = sqlS[sqlpt-3 : sqlpt+1]
 		//line sql.y:3482
 		{
-			sqlVAL.union.val = sqlDollar[3].union
+			sqlVAL.union.val = sqlDollar[3].union.durationField()
 		}
 	case 568:
 		sqlDollar = sqlS[sqlpt-3 : sqlpt+1]
@@ -8505,13 +8505,13 @@ sqldefault:
 		sqlDollar = sqlS[sqlpt-3 : sqlpt+1]
 		//line sql.y:3490
 		{
-			sqlVAL.union.val = sqlDollar[3].union
+			sqlVAL.union.val = sqlDollar[3].union.durationField()
 		}
 	case 570:
 		sqlDollar = sqlS[sqlpt-3 : sqlpt+1]
 		//line sql.y:3494
 		{
-			sqlVAL.union.val = sqlDollar[3].union
+			sqlVAL.union.val = sqlDollar[3].union.durationField()
 		}
 	case 571:
 		sqlDollar = sqlS[sqlpt-0 : sqlpt+1]

--- a/pkg/sql/parser/sql.y
+++ b/pkg/sql/parser/sql.y
@@ -3466,7 +3466,7 @@ opt_interval:
   }
 | interval_second
   {
-    $$.val = $1
+    $$.val = $1.durationField()
   }
 // Like Postgres, we ignore the left duration field. See explanation:
 // https://www.postgresql.org/message-id/20110510040219.GD5617%40tornado.gateway.2wire.net
@@ -3484,7 +3484,7 @@ opt_interval:
   }
 | DAY TO interval_second
   {
-    $$.val = $3
+    $$.val = $3.durationField()
   }
 | HOUR TO MINUTE
   {
@@ -3492,11 +3492,11 @@ opt_interval:
   }
 | HOUR TO interval_second
   {
-    $$.val = $3
+    $$.val = $3.durationField()
   }
 | MINUTE TO interval_second
   {
-    $$.val = $3
+    $$.val = $3.durationField()
   }
 | /* EMPTY */ {}
 

--- a/pkg/sql/testdata/datetime
+++ b/pkg/sql/testdata/datetime
@@ -687,6 +687,22 @@ SHOW TIME ZONE
 -5
 
 statement ok
+SET TIME ZONE INTERVAL '+04:00' HOUR TO MINUTE;
+
+query T
+SELECT '2015-08-24 21:45:45.53453'::timestamptz
+----
+2015-08-24 21:45:45.53453 +0400 +0400
+
+statement ok
+SET TIME ZONE INTERVAL '-04:00' MINUTE TO SECOND;
+
+query T
+SELECT '2015-08-24 21:45:45.53453'::timestamptz
+----
+2015-08-24 21:45:45.53453 -0400 -0400
+
+statement ok
 SET TIME ZONE 0
 
 query T

